crates/circuit/src/parameter_table.rs

```diff
@@ -10,164 +10,395 @@
-        self.index_ids.contains(&key)
+    /// Lookup the Python parameter object by name.
+    pub fn py_parameter_by_name(&self, name: &PyBackedStr) -> Option<&Py<PyAny>> {
+        self.by_name
+            .get(name)
+            .map(|uuid| &self.by_uuid[uuid].object)
     }
 
-    pub fn __iter__(&self) -> ParamEntryKeys {
-        ParamEntryKeys {
-            keys: self.index_ids.iter().copied().collect(),
-            iter_pos: 0,
+    /// Get the (maybe cached) Python list of the sorted `Parameter` objects.
+    pub fn py_parameters<'py>(&mut self, py: Python<'py>) -> Bound<'py, PyList> {
+        if let Some(py_parameters) = self.py_parameters.as_ref() {
+            return py_parameters.clone_ref(py).into_bound(py);
         }
+        self.ensure_sorted();
+        let out = PyList::new_bound(
+            py,
+            self.order
+                .iter()
+                .map(|uuid| self.by_uuid[uuid].object.clone_ref(py).into_bound(py)),
+        );
+        self.py_parameters = Some(out.clone().unbind());
+        out
     }
-}
 
-#[derive(Clone, Debug)]
-#[pyclass(freelist = 20, module = "qiskit._accelerate.circuit")]
-pub(crate) struct ParamTable {
-    /// Mapping of parameter uuid (as an int) to the Parameter Entry
-    pub table: HashMap<u128, ParamEntry>,
-    /// Mapping of parameter name to uuid as an int
-    pub names: HashMap<String, u128>,
-    /// Mapping of uuid to a parameter object
-    pub uuid_map: HashMap<u128, PyObject>,
-}
+    /// Get a Python set of all tracked `Parameter` objects.
+    pub fn py_parameters_unsorted<'py>(&self, py: Python<'py>) -> PyResult<Bound<'py, PySet>> {
+        PySet::new_bound(py, self.by_uuid.values().map(|info| &info.object))
+    }
 
-impl ParamTable {
-    pub fn insert(&mut self, py: Python, parameter: PyObject, entry: ParamEntry) -> PyResult<()> {
-        let uuid: u128 = parameter
-            .getattr(py, intern!(py, "_uuid"))?
-            .getattr(py, intern!(py, "int"))?
-            .extract(py)?;
-        let name: String = parameter.getattr(py, intern!(py, "name"))?.extract(py)?;
-
-        if self.names.contains_key(&name) && !self.table.contains_key(&uuid) {
-            return Err(CircuitError::new_err(format!(
-                "Name conflict on adding parameter: {}",
-                name
-            )));
+    /// Ensure that the `order` field is populated and sorted.
+    fn ensure_sorted(&mut self) {
+        // If `order` is already populated, it's sorted; it's the responsibility of the methods of
+        // this struct that mutate it to invalidate the cache.
+        if !self.order.is_empty() {
+            return;
         }
-        self.table.insert(uuid, entry);
-        self.names.insert(name, uuid);
-        self.uuid_map.insert(uuid, parameter);
+        self.order.reserve(self.by_uuid.len());
+        self.order.extend(self.by_uuid.keys());
+        self.order.sort_unstable_by_key(|uuid| {
```

Comment:
I wonder if there is a threshold where rayon's `par_sort_unstable_by_key()` would be useful here. We can look at that in the future.

Reply:
Quite possibly yes, unless it interferes with the fastest paths for data that's already sorted.  The cases we need this to have the highest performance correspond to IBM backends' fast parametric updates, where all the parameters are likely to be in the circuit in order to begin with.